### PR TITLE
Cleanup registered handlers on destroy

### LINKED ISSUE #13: Teardown registered handlers when service is destroyed.

### PR DESCRIPTION
**Pull request by jasonmit**
Fixes #13 and #12

### CODE REVIEW

File: addon/services/unified-event-handler.js [hunk 1]
```diff
@@ -40,8 +40,11 @@ const generateId = (function() {
 }());
 
 export default Ember.Service.extend(Ember.Evented, {
-  // Keeps track of the handlers that have been registered
-  [_HANDLER_MAP]: {},
+  init() {
+    this._super(...arguments);
+
+    this[_HANDLER_MAP] = {};
```
Comment (trentmwillis) [inline addon/services/unified-event-handler.js]: Let's update this to `Object.create(null)`. We really just want an empty dictionary.

File: tests/unit/services/unified-event-handler-test.js [hunk 2]
```diff
@@ -28,6 +28,20 @@ test('register binds callback to the event on the specified target', function(as
   service.unregister('window', 'scroll', callbackStub);
 });
 
+test('unregisters event listeners when service is destroyed', function(assert) {
+  let done = assert.async();
+  let callbackStub = sandbox.stub();
+  service.register('window', 'scroll', callbackStub);
```
Comment (trentmwillis) [inline tests/unit/services/unified-event-handler-test.js]: Let's update this test to have multiple things registered, to make sure we're unregistering all the handlers.

File: addon/services/unified-event-handler.js [hunk 3]
```diff
@@ -145,6 +148,24 @@ export default Ember.Service.extend(Ember.Evented, {
     this.on(domHandlerInfo.emberEventName, callback);
   },
 
+  willDestroy() {
+    let handlerMap = this[_HANDLER_MAP];
+
+    for (let target in handlerMap) {
+      if (handlerMap.hasOwnProperty(target)) {
```
Comment (trentmwillis) [inline addon/services/unified-event-handler.js]: With the change above to `Object.create(null)`, we shouldn't need this check anymore.

Can we do a similar thing for the nested objects as well?

File: addon/services/unified-event-handler.js [hunk 4]
```diff
@@ -145,6 +148,24 @@ export default Ember.Service.extend(Ember.Evented, {
...
+        let handlerTarget = handlerMap[target];
+        for (let eventName in handlerTarget) {
+          if (handlerTarget.hasOwnProperty(eventName)) {
+            let handlerInfo = handlerTarget[eventName];
+            for (let i=0;i<handlerInfo.emberHandlers.length;i++) {
```
Comment (trentmwillis) [inline addon/services/unified-event-handler.js]: Nitpick, but can you add a space after each of the `;`?

**Comment by jasonmit** [comment]
@trentmwillis all feedback addressed.

File: addon/services/unified-event-handler.js [hunk 5]
```diff
@@ -145,6 +147,21 @@ export default Ember.Service.extend(Ember.Evented, {
...
   },
 
+  willDestroy() {
+    let handlerMap = this[_HANDLER_MAP];
+
+    for (let target in handlerMap) {
+      let handlerTarget = handlerMap[target];
+      for (let eventName in handlerTarget) {
+        let handlerInfo = handlerTarget[eventName];
+        let emberHandlers = handlerInfo.emberHandlers.slice(0);
```
Comment (jasonmit) [inline addon/services/unified-event-handler.js]: Clone array since unregister mutates handlerInfo.emberHandlers as we enumerate it.

**Comment by jasonmit** [comment]
@trentmwillis anything needed here on my end? Would like to get off my branch if possible.

**Comment by trentmwillis** [comment]
Sorry for missing this. Changes look good to me. I'll publish a new release later today.

Thanks for the contribution!

**Pull request merged**